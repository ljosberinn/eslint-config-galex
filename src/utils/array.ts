export const uniqueArrayEntries = <T>(arr: T[]): T[] => [...new Set(arr)];
