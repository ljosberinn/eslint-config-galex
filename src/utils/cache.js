const cache = {
  createdAt: null,
  config: null,
  dependencies: null,
};

const mustInvalidate = ({ now, dependencies }) => {
  // neither enabled nor previously ran
  if (!dependencies.cacheOptions.enabled || !cache.config) {
    return true;
  }

  const hasChangedDependencies =
    JSON.stringify(dependencies) !== cache.dependencies;

  if (hasChangedDependencies) {
    return true;
  }

  return now - dependencies.cacheOptions.expiresAfterMs > cache.createdAt;
};

const set = ({ now, config, dependencies }) => {
  cache.createdAt = now;
  cache.config = config;
  cache.dependencies = JSON.stringify(dependencies);
};

/**
 * @returns {object | null}
 */
const get = () => cache.config;

module.exports = {
  set,
  mustInvalidate,
  get,
};
