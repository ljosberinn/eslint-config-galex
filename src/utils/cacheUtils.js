const set = (cache, { now, config, dependencies }) => {
  cache.createdAt = now;
  cache.config = config;
  cache.dependencies = dependencies;
};

const mustInvalidate = (cache, { now, dependencies }) => {
  // neither enabled nor previously ran
  if (!dependencies.cacheOptions.enabled || !cache.config) {
    return true;
  }

  const hasChangedDependencies =
    JSON.stringify(dependencies) !== JSON.stringify(cache.dependencies);

  if (hasChangedDependencies) {
    return true;
  }

  return now - dependencies.cacheOptions.expiresAfterMs > cache.createdAt;
};

const createCache = () => {
  const cache = {
    createdAt: null,
    config: null,
    dependencies: null,
  };

  return {
    cache,
    set,
    mustInvalidate,
  };
};

module.exports = {
  createCache,
};
