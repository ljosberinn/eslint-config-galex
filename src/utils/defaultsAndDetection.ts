import type { Linter } from 'eslint';

export const plugins = ['import', 'unicorn', 'promise', 'sonarjs'];
export const reactFlavours = ['react', 'preact', 'next', '@remix-run/react'];

/**
 * @see https://www.npmjs.com/org/testing-library
 */
export const testingLibFamily = [
  'angular',
  'cypress',
  'dom',
  // jasmine-dom
  // jest-dom
  // jest-native
  'nightwatch',
  'preact',
  'preact-hooks',
  'react',
  'react-hooks',
  'react-native',
  'svelte',
  'testcafe',
  // user-event
  'vue',
];

export const parserOptions: Required<Linter.ParserOptions> = {
  ecmaVersion: 'latest',
  sourceType: 'module',
  ecmaFeatures: {},
};
