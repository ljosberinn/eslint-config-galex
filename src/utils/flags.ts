import type { Linter } from 'eslint';

import type { ESLintConfig, Flags } from '../types';

const severityLevels: Record<
  Exclude<Linter.RuleLevel, Linter.Severity>,
  Linter.Severity
> = {
  off: 0,
  warn: 1,
  error: 2,
};

const convertRuleToEslintInternalValue: FlagProcessor = options => {
  if (Array.isArray(options)) {
    const [ruleLevel, ...rest] = options;

    if (typeof ruleLevel === 'number') {
      return options;
    }

    return [severityLevels[ruleLevel], ...rest];
  }

  return typeof options === 'number' ? options : severityLevels[options];
};

const incrementalAdoptionRuleDowngrade: FlagProcessor = value => {
  const valueToForward =
    typeof value === 'number' || typeof value === 'string' ? value : value[0];

  if (typeof valueToForward === 'string') {
    if (valueToForward === 'error') {
      return 'warn';
    }

    return 'off';
  }

  return valueToForward === 2 ? 1 : 0;
};

const blankSlateDowngrade: FlagProcessor = () => {
  return 'off';
};

type FlagProcessor = (value: Linter.RuleEntry) => Linter.RuleEntry;

const createFlagFilterArray = ({
  convertToESLintInternals,
  incrementalAdoption,
  blankSlate,
}: Required<Flags>): FlagProcessor[] => {
  return [
    convertToESLintInternals && convertRuleToEslintInternalValue,
    incrementalAdoption && incrementalAdoptionRuleDowngrade,
    blankSlate && blankSlateDowngrade,
  ].filter((fn): fn is FlagProcessor => fn !== false);
};

export const applyFlags = (
  rules: Linter.RulesRecord,
  flags: Flags
): ESLintConfig['rules'] => {
  const convertToESLintInternals = flags.convertToESLintInternals ?? false;
  const incrementalAdoption = flags.incrementalAdoption ?? false;
  const blankSlate = flags.blankSlate ?? false;

  const hasFlags =
    convertToESLintInternals || incrementalAdoption || blankSlate;

  if (!hasFlags) {
    return rules;
  }

  const fnsToApply = createFlagFilterArray({
    blankSlate,
    convertToESLintInternals,
    incrementalAdoption,
  });

  return Object.fromEntries(
    Object.entries(rules).map(([ruleName, options]) => [
      ruleName,
      fnsToApply.reduce((acc, fn) => fn(acc), options),
    ])
  );
};
