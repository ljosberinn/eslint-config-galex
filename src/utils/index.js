const {
  jestOverrideType,
  reactOverrideType,
  tsOverrideType,
  storybookOverrideType,
} = require('./overrideTypes');

const severityLevels = {
  off: 0,
  warn: 1,
  error: 2,
};

const convertRuleToEslintInternalValue = value => {
  if (severityLevels[value] !== undefined) {
    return severityLevels[value];
  }

  if (Array.isArray(value) && severityLevels[value[0]] !== undefined) {
    const options = [...value];
    options[0] = severityLevels[options[0]];

    return options;
  }

  return value;
};

/**
 * naive version detection
 *
 * @param {string|number} version
 * @param {{
 *  major: number;
 *  minor?: number;
 *  patch?: number;
 * }} minRequiredVersion
 */
const fulfillsVersionRequirement = (
  version,
  { major, minor = 0, patch = 0 }
) => {
  try {
    const [depMajor, depMinor, depPatch] = version
      .split('.')
      .map(str => Number.parseInt(str));

    // version is identical to required
    if (depMajor === major && depMinor === minor && depPatch === patch) {
      return true;
    }

    // bail if major is higher than required
    if (depMajor > major) {
      return true;
    }

    // bail if major is lower than required
    if (depMajor < major) {
      return false;
    }

    // major is equal to requirements

    // bail if minor is higher than required
    if (depMinor > minor) {
      return true;
    }

    // bail if minor is lower than required
    if (depMinor < minor) {
      return false;
    }

    // major and minor are equal to requirements

    // bail if patch is higher than required
    if (depPatch > patch) {
      return true;
    }

    // bail if patch is lower than required
    return false;
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error(`Error parsing version: ${error.message}`);

    return false;
  }
};

const applyFlagFilter = (rules, { convertToESLintInternals = true }) => {
  if (!convertToESLintInternals) {
    return rules;
  }

  return Object.fromEntries(
    Object.entries(rules).map(([key, value]) => [
      key,
      convertRuleToEslintInternalValue(value),
    ])
  );
};

const pseudoDeepMerge = (override, previous) =>
  Object.entries(override).reduce((carry, [key, value]) => {
    /* istanbul ignore next line 217 is supposedly uncovered :shrug: */
    if (carry[key]) {
      if (Array.isArray(carry[key])) {
        return {
          ...carry,
          [key]: [
            ...new Set([
              ...carry[key],
              ...(Array.isArray(value) ? value : [value]),
            ]),
          ],
        };
      }

      if (typeof carry[key] === 'object') {
        return {
          ...carry,
          [key]: {
            ...carry[key],
            ...value,
          },
        };
      }

      carry[key] = value;
    }

    return carry;
  }, previous);

const overrideOrder = {
  [jestOverrideType]: 0,
  [tsOverrideType]: 1,
  [reactOverrideType]: 2,
  [storybookOverrideType]: 3,
};

/**
 * @param {object[]} overrides
 */
const mergeSortOverrides = overrides => {
  return overrides
    .filter(Boolean)
    .reduce((carry, override) => {
      const isInternalOverride = !!override.overrideType;

      if (isInternalOverride) {
        const previousDefaultOverrideTypeIndex = carry.findIndex(
          o => o.overrideType === override.overrideType
        );

        if (previousDefaultOverrideTypeIndex > -1) {
          return carry.map((dataset, index) =>
            index === previousDefaultOverrideTypeIndex
              ? pseudoDeepMerge(override, dataset)
              : dataset
          );
        }
      }

      return [...carry, override];
    }, [])
    .sort((a, b) => {
      const priorityA = a.overrideType ? overrideOrder[a.overrideType] : -1;
      const priorityB = b.overrideType ? overrideOrder[b.overrideType] : -1;

      return priorityB - priorityA;
    });
};

module.exports = {
  applyFlagFilter,
  fulfillsVersionRequirement,
  mergeSortOverrides,
};
