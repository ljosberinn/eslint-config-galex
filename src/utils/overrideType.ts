import { OverrideESLintConfig, WithOverrideType } from '../types';

const createOverrideType = (type: string) => `eslint-config-galex/${type}`;

export const jestOverrideType = createOverrideType('jest');
export const jestConfigOverrideType = createOverrideType('jest-config');
export const reactOverrideType = createOverrideType('react');
export const tsOverrideType = createOverrideType('typescript');
export const storybookOverrideType = createOverrideType('storybook');

export const dropOverrideType = (
  override: WithOverrideType<OverrideESLintConfig>
): OverrideESLintConfig => {
  const { overrideType, ...rest } = override;

  return rest;
};
