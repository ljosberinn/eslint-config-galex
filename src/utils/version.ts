const cache = new Map<string, [number, number, number]>();

const toNumbers = (str: string): [number, number, number] => {
  const maybeCached = cache.get(str);

  if (maybeCached) {
    return maybeCached;
  }

  /* istanbul ignore next major cant ever be undefined, even if only "." is forwarded */
  const [major = 0, minor = 0, patch = 0] = str
    .split('.')
    .map(str =>
      [...str]
        .map(char => Number.parseInt(char))
        .filter(maybeNumber => !Number.isNaN(maybeNumber))
        .join('')
    )
    .map(char => {
      const maybeNumber = Number.parseInt(char);

      return typeof maybeNumber !== 'number' || Number.isNaN(maybeNumber)
        ? 0
        : maybeNumber;
    });

  const sane: [number, number, number] = [major, minor, patch];

  cache.set(str, sane);

  return sane;
};

export const fulfillsVersionRequirement = ({
  given,
  expected,
}: {
  expected: string;
  given: string;
}): boolean => {
  const [givenMajor, givenMinor, givenPatch] = toNumbers(given);
  const [expectedMajor, expectedMinor, expectedPatch] = toNumbers(expected);

  // satisfying major
  if (givenMajor > expectedMajor) {
    return true;
  }

  if (givenMajor === expectedMajor) {
    // satisfying minor
    if (givenMinor > expectedMinor) {
      return true;
    }

    // satisfying patch
    return givenMinor === expectedMinor && givenPatch >= expectedPatch;
  }

  return false;
};
