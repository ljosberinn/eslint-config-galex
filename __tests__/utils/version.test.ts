import { fulfillsVersionRequirement } from '../../src/utils/version';

test.each([
  ['4.1.2', '^3.7.0', true],
  ['4.1.2', '^4.1.2', true],
  ['4.1.2', '^4.2.0', false],
  ['3.0.0', '4.1.2', false],
  ['3.0', '4.1.2', false],
  ['3', '4.1.2', false],
  ['.', '4.1.2', false],
])('fulfillsVersionRequirement(%s, %s)', (version, required, result) => {
  expect(
    fulfillsVersionRequirement({ given: version, expected: required })
  ).toBe(result);
  expect(
    fulfillsVersionRequirement({ given: version, expected: required })
  ).toMatchSnapshot();
});
