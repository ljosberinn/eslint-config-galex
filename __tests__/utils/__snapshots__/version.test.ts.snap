// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fulfillsVersionRequirement(., 4.1.2) 1`] = `false`;

exports[`fulfillsVersionRequirement(3, 4.1.2) 1`] = `false`;

exports[`fulfillsVersionRequirement(3.0, 4.1.2) 1`] = `false`;

exports[`fulfillsVersionRequirement(3.0.0, 4.1.2) 1`] = `false`;

exports[`fulfillsVersionRequirement(4.1.2, ^3.7.0) 1`] = `true`;

exports[`fulfillsVersionRequirement(4.1.2, ^4.1.2) 1`] = `true`;

exports[`fulfillsVersionRequirement(4.1.2, ^4.2.0) 1`] = `false`;
